File: src/services/audit.ts
import type { AuditRecord } from "../core/models.js";
import type { AuditRepository } from "../db/auditRepository.js";
import { WriteQueue } from "../db/writeQueue.js";

export interface AuditLog {
  append(record: AuditRecord): void;
  findByGroup(groupId: string): AuditRecord[];
  all(): AuditRecord[];
  /** 持久化实现才需要；内存实现可省略。 */
  flush?(): Promise<void>;
}

/**
 * 审计日志存储。
 *
 * 注入仓储后升级为持久化：启动时 `load()` 全量载入，`append()` 同步写入内存缓存、
 * 异步写穿透到数据库；未注入仓储时退化为纯内存实现，便于单元测试。
 */
export class AuditLogStore implements AuditLog {
  private readonly records: AuditRecord[] = [];
  private readonly repository: AuditRepository | undefined;
  private readonly queue: WriteQueue | undefined;

  public constructor(repository?: AuditRepository, queue?: WriteQueue) {
    this.repository = repository;
    this.queue = repository ? (queue ?? new WriteQueue()) : undefined;
  }

  public get persistent(): boolean {
    return this.repository !== undefined;
  }

  public async load(): Promise<void> {
    if (!this.repository) {
      return;
    }
    const records = await this.repository.findAll();
    this.records.length = 0;
    this.records.push(...records);
  }

  public async flush(): Promise<void> {
    await this.queue?.flush();
  }

  public append(record: AuditRecord): void {
    this.records.push(record);
    const repository = this.repository;
    if (repository) {
      this.queue?.enqueue("audit.append", () => repository.append(record));
    }
  }

  public findByGroup(groupId: string): AuditRecord[] {
    return this.records.filter((record) => record.groupId === groupId);
  }

  public all(): AuditRecord[] {
    return [...this.records];
  }
}

File: src/qq_group_guard/services/audit.py
"""审计日志服务。

Phase 1 会接入 PostgreSQL；当前提供内存实现用于测试和本地开发。
"""

from __future__ import annotations

from qq_group_guard.core.models import AuditRecord


class AuditLog:
    """审计日志抽象。"""

    def append(self, record: AuditRecord) -> None:
        raise NotImplementedError

    def find_by_group(self, group_id: str) -> tuple[AuditRecord, ...]:
        raise NotImplementedError


class InMemoryAuditLog(AuditLog):
    """内存审计日志，仅用于测试和本地开发。"""

    def __init__(self) -> None:
        self._records: list[AuditRecord] = []

    def append(self, record: AuditRecord) -> None:
        self._records.append(record)

    def find_by_group(self, group_id: str) -> tuple[AuditRecord, ...]:
        return tuple(record for record in self._records if record.group_id == group_id)

    def all(self) -> tuple[AuditRecord, ...]:
        return tuple(self._records)
